src/crab/abs_transformer.hpp

```diff
@@ -0,0 +1,1044 @@
+    void bitwise_or(variable_t lhs, variable_t op2) { apply(m_inv, crab::bitwise_binop_t::OR, lhs, lhs, op2); }
+    void bitwise_or(variable_t lhs, number_t op2) { apply(m_inv, crab::bitwise_binop_t::OR, lhs, lhs, op2); }
+    void bitwise_xor(variable_t lhs, variable_t op2) { apply(m_inv, crab::bitwise_binop_t::XOR, lhs, lhs, op2); }
+    void bitwise_xor(variable_t lhs, number_t op2) { apply(m_inv, crab::bitwise_binop_t::XOR, lhs, lhs, op2); }
+    void shl_overflow(variable_t lhs, variable_t op2) { apply(m_inv, crab::bitwise_binop_t::SHL, lhs, lhs, op2, true); }
+    void shl_overflow(variable_t lhs, number_t op2) { apply(m_inv, crab::bitwise_binop_t::SHL, lhs, lhs, op2, true); }
+    void lshr(variable_t lhs, variable_t op2) { apply(m_inv, crab::bitwise_binop_t::LSHR, lhs, lhs, op2); }
+    void lshr(variable_t lhs, number_t op2) { apply(m_inv, crab::bitwise_binop_t::LSHR, lhs, lhs, op2); }
+    void ashr(variable_t lhs, variable_t op2) { apply(m_inv, crab::bitwise_binop_t::ASHR, lhs, lhs, op2); }
+    void ashr(variable_t lhs, number_t op2) { apply(m_inv, crab::bitwise_binop_t::ASHR, lhs, lhs, op2); }
+
+  protected:
+    void assume(const linear_constraint_t& cst) { assume(m_inv, cst); }
+    void assume(AbsDomain& inv, const linear_constraint_t& cst) { inv += cst; }
+
+    void require(const linear_constraint_t& cst, std::string s) { require(this->m_inv, cst, s); }
+    virtual void require(AbsDomain& inv, const linear_constraint_t& cst, std::string s) { assume(inv, cst); }
+
+    void havoc(variable_t v) { m_inv -= v; }
+    void assign(variable_t lhs, variable_t rhs) { m_inv.assign(lhs, rhs); }
+    void assign(variable_t lhs, number_t rhs) { m_inv.assign(lhs, rhs); }
+
+    void no_pointer(int i) {
+        assign(reg_type(i), T_NUM);
+        havoc(reg_offset(i));
+    };
+    void no_pointer(Reg r) { no_pointer(r.v); }
+
+    static linear_constraint_t is_shared(variable_t v) {
+        using namespace dsl_syntax;
+        return v > T_SHARED;
+    }
+
+    static linear_constraint_t is_pointer(Reg v) {
+        using namespace dsl_syntax;
+        return reg_type(v) >= T_CTX;
+    }
+    static linear_constraint_t is_init(Reg v) {
+        using namespace dsl_syntax;
+        return reg_type(v) > T_UNINIT;
+    }
+    static linear_constraint_t is_shared(Reg v) { return is_shared(reg_type(v)); }
+    static linear_constraint_t is_not_num(Reg v) {
+        using namespace dsl_syntax;
+        return reg_type(v) > T_NUM;
+    }
+
+    void overflow(variable_t lhs) {
+        using namespace dsl_syntax;
+        auto interval = m_inv[lhs];
+        // handle overflow, assuming 64 bit
+        number_t max(std::numeric_limits<int64_t>::max() / 2);
+        number_t min(std::numeric_limits<int64_t>::min() / 2);
+        if (interval.lb() <= min || interval.ub() >= max)
+            havoc(lhs);
+    }
+
+  public:
+    intra_abs_transformer(const AbsDomain& inv) : m_inv(inv) {}
+
+    void operator()(Assume const& s) {
```

Comment:
It would help some comments describing what the operation does. If both operands are registers then either they have same types or they have different types. If they have different types then one must be a null pointer etc